### PR TITLE
Adding ws0010 and winstar_weh drivers

### LINKED ISSUE #294: Support for Winstar WS0010 driver
I recently purchased a [Winstar WEG010016A display](https://www.winstar.com.tw/products/oled-module/graphic-oled-display/weg010016a.html), incorrectly thinking that there would be ample documentation on how to use it online. I've spent several hours trawling through the internet to find if others have used it before with a raspberry pi and I managed to find a few references to different winstar _character_ displays being used, but none of this model, and none even of graphical winstar oled screens controlled by the WS0010 universal driver.

I did manage to find [this pull request](https://github.com/rm-hull/luma.core/pull/180) from 14 days ago which seems to add support for Winstar WEG displays to the luma.core, but as far as I can see (I'm quite a novice to SPI and displays in general), that support hasn't yet been added to the luma.oled project.

I have used this library before for a SSD1322 oled display, but for my current project I needed a red on black oled, and the above model is the only one I could find.

My question is, am I able to use the luma.oled library with my Winstar WEG display, and if not, it would be great if support could be added. I'm aware I sound a bit arrogant but as much as I'd like to contribute myself, I am really not sure where to even begin; the datasheets for these displays send my head spinning :/

### PR DESCRIPTION
**Pull request by dhrone**
Added ws0010 driver (and related winstar_weh) driver.  Should resolve #294

### CODE REVIEW

**Comment by GRA0007** [comment]
I have tried testing this branch with my WEG010016A display, but I couldn't get it to work. I'm not 100% sure I'm wiring it up correctly, or maybe there's an issue with the code.

This is the python I'm using:
```py
from luma.core.interface.serial import parallel
from luma.core.render import canvas
from luma.oled.device import ws0010

p = parallel(RS=7, E=8, PINS=[25,24,23,27])
d = ws0010(p)
d.text = 'HELLO'
```

I'm wiring up my display like the following:
WEG010016A | Raspberry Pi
---|---
Pin 1 (GND) | GND
Pin 2 (5V) | 5V
Pin 4 (RS) | GPIO 7
Pin 5 (R/W) | GND
Pin 6 (E) | GPIO 8
Pin 7 (DB0) | GPIO 25
Pin 8 (DB1) | GPIO 24
Pin 9 (DB2) | GPIO 23
Pin 10 (DB3) | GPIO 27

I'm assuming I don't have to use all 8 data bus pins on the display, but I can't find any documentation anywhere suggesting one way or another.

File: luma/oled/device/__init__.py [hunk 1]
```diff
@@ -743,3 +746,232 @@ def _set_position(self, top, right, bottom, left):
...
+        return self.font.load(ft)
+
+
+class winstar_weh(ws0010):
+    """
+    Serial interface to a monochrome Winstar WEH OLED display.  This is the
+    character version of the display using the ws0010 controller.  This class
+    provides the same ``text`` property as the ws0010 interface so you can set
+    the text value which will be rendered to the display's screen.  This
+    interface uses a varient of the ws0010 controller's built-in font that is
```
Comment (rm-hull) [inline luma/oled/device/__init__.py:929]: Spelling: variant

File: luma/oled/device/__init__.py [hunk 2]
```diff
@@ -743,3 +746,232 @@ def _set_position(self, top, right, bottom, left):
...
+        OLED display.
+        """
+        assert(image.mode == self.mode)
+        assert(image.size == self.size)
+
+        image = self.preprocess(image)
+
+        if self.framebuffer.redraw_required(image):
+            # Expand bounding box to align to cell height boundary (8)
+            # Should consider whether this should be moved into framebuffer class
```
Comment (rm-hull) [inline luma/oled/device/__init__.py:885]: Prefix comments like these with either TODO or FIXME so that the ide highlights them for future attention

**Review by rm-hull — CHANGES_REQUESTED** [review]
Again, blown away with attention to detail in this PR. Great stuff !
Could you also change the `setup.py` file to enforce the min required version of luma.core to be >= 1.16.0 too please?

**Comment by rm-hull** [comment]
@GRA0007 could you be more specific? Did it just not display anything or did you see an error?

Maybe try putting a `time.sleep(10000)` at the end? The display driver will shutdown And cleanup the display once a program exits

**Comment by GRA0007** [comment]
@rm-hull Sorry, yeah it simply doesn't display anything at all. I did try sleeping after, and I tried just in the live python interpreter too to no avail.

**Comment by rm-hull** [comment]
@GRA0007 will the device run on 3.3V? reason i ask is because the GPIOs are all at 3.3V levels but you are supplying power at 5.0V, and it may be that you might need to use a voltage level shifter. I haven't looked at any tech-specs for these devices so i dont know for sure.

Actually that reminds me, @dhrone could you add any spec PDF's for the devices to the docs (https://github.com/rm-hull/luma.oled/tree/master/doc/tech-spec) and make references to them in the intro docs (see https://github.com/rm-hull/luma.oled/blob/master/doc/intro.rst). It would also be good to elaborate on the hardware setup too for the parallel interface (see https://github.com/rm-hull/luma.oled/blob/master/doc/intro.rst)

**Comment by dhrone** [comment]
@GRA0007  Very close but your wiring is off.  When in 4 bit mode which is the default, the four PINS should be attached to data lines db4-db7 not db0-db3.  @rm-hull this display requires 5Vs.  There is a good document on the ws0010 at [pixace.com](https://picaxe.com/docs/oled.pdf) and the pinout data is at the [winstar site](https://www.winstar.com.tw/products/oled-module/graphic-oled-display/weg010016a.html).    @GRA0007, this the wiring diagram you should use...

WEG010016A | Raspberry Pi
-- | --
Pin 1 (GND) | GND
Pin 2 (5V) | 5V
Pin 4 (RS) | GPIO 7
Pin 5 (R/W) | GND
Pin 6 (E) | GPIO 8
Pin 11 (DB4) | GPIO 25
Pin 12 (DB5) | GPIO 24
Pin 13 (DB6) | GPIO 23
Pin 14 (DB7) | GPIO 27

**Comment by GRA0007** [comment]
@dhrone omg thank you so much! Using pins 4-7 is a bit unintuitive, and I couldn't find that documentation anywhere. It's working perfectly now. I'll keep testing and report back if there are any issues, but it looks great. Thanks again :D

**Comment by dhrone** [comment]
@GRA0007 glad to hear and somewhat relieved.  These parallel-bus devices can be difficult to debug when there are issues...

One feature you may want to play with is the ability to combine fonts to produce a font that contains symbols which are missing from the one you have selected.  The combine method is part of the embedded_font class.  The ws0010.font instance variable is an embedded_font object.  So you can combine the font tables (or other fonts) to form versions that fill in for missing symbols.  For examples...
```
  from time import sleep
  p = parallel(RS=7, E=8, PINS=[25,24,23,27])
  d = ws0010(p, selected_font='FT01')
  d.text = '\u25b6 Play'
  ft11fnt = d.get_font('FT11')
  d.font.combine(ft11fnt, '\u25b6')
  sleep(2)
  d.text = '\u25b6 Play'
```
 In the first d.text command, the Unicode symbol u25b6 (Black right-pointing triangle) does not exist in the selected font (FT00) so the undefined character is displayed instead.  But once the selected font has been combined with the FT11 font it can display the character.  Note, the second parameter to the combine method allows you to select which characters to move from the source font.  If you leave this out, every character (at least the ones with unicode values mapped to them) will be moved.

File: luma/oled/const.py [hunk 3]
```diff
@@ -39,3 +39,552 @@ class ssd1362(common):
     DISPLAYON = 0xAF
     DISPLAYOFF = 0xAE
     SETCONTRAST = 0x81
+
+
+class ws0010(object):
+    CLEAR = 0x01
```
Comment (thijstriemstra) [inline luma/oled/const.py:45]: add version marker for new classes (and description in docstring).

File: luma/oled/const.py [hunk 4]
```diff
@@ -39,3 +39,552 @@ class ssd1362(common):
...
+    POWERON = 0x17
+    GRAPHIC = 0x08
+    CHAR = 0x00
+    FUNCTIONSET = 0x29
+    DL8 = 0x10
+    DL4 = 0x00
+    DDRAMADDR = 0x80
+    CGRAMADDR = 0x40
+    FONTDATA = {
+        # TODO: Extend and verify Unicode Mappings
```
Comment (thijstriemstra) [inline luma/oled/const.py:60]: if it's a todo not happening in this pull request, please open a ticket and reference it here

File: luma/oled/const.py [hunk 5]
```diff
@@ -39,3 +39,552 @@ class ssd1362(common):
...
+    # Replicate FT11 data to support making 5x10 version
+    FONTDATA['mappings'].append(FONTDATA['mappings'][3])
+    FONTDATA['fonts'].append(FONTDATA['fonts'][3])
+    FONTDATA['metrics'].append(dict(FONTDATA['metrics'][3]))
+    FONTDATA['metrics'][7]['name'] = 'FT11_5x10'
+    FONTDATA['metrics'][7]['glyph_size'] = (5, 10)
+
+
+class winstar_weh(ws0010):
+    FONTDATA = {
```
Comment (thijstriemstra) [inline luma/oled/const.py:582]: add version marker

File: tests/test_ws0010.py [hunk 6]
```diff
@@ -0,0 +1,146 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+# Copyright (c) 2013-17 Richard Hull and contributors
```
Comment (thijstriemstra) [inline tests/test_ws0010.py:3]: should be 2020

File: tests/test_ws0010.py [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+# See LICENSE.rst for details.
+
+"""
+Tests for the :py:class:`luma.lcd.device.ws0010` device.
```
Comment (thijstriemstra) [inline tests/test_ws0010.py:7]: there's a typo "lcd"

**Review by thijstriemstra — CHANGES_REQUESTED** [review]

**Comment by dhrone** [comment]
@GRA0007  I was just reading up on the specifications for the weg010016 and there is some ambiguity in whether it will run with 3.3v or requires 5v.  Have you ever attempted to run it at 3.3v?  If it will run at 3.3v that will be safer for the RPi as the docs suggest that the voltage on the data pins can be as high as VDD so if we're supplying 5Vs we may get that back.  Having said that though, I've been running these displays at 5Vs on RPis for a long time with no ill effects.  That might relate to the fact that I only use them in write-mode with RS held low by tying it to ground.

**Review by thijstriemstra — COMMENTED** [review]
It's also missing some documentation, maybe a note in `doc/hardware.rst`? And the readme should mention the new devices, as well as adding keywords to setup.py.

**Review by thijstriemstra — APPROVED** [review]
great, thanks!

**Comment by thijstriemstra** [comment]
ps. @rm-hull can you invite @dhrone to the github team/collaborators?

File: luma/oled/device/__init__.py (hunk 1 above)
Comment (dhrone) [reply luma/oled/device/__init__.py]: Resolved

File: luma/oled/device/__init__.py (hunk 2 above)
Comment (dhrone) [reply luma/oled/device/__init__.py]: Added TODO to the comment.  I think it is probably ok as-is though so let me know if you want me to open an issue to address it, or remove the comment.  It was really just me capturing my thought process as I was coding the routine.

File: setup.py [hunk 8]
```diff
@@ -64,7 +65,7 @@ def find_version(*file_paths):
     packages=find_packages(),
     namespace_packages=["luma"],
     zip_safe=False,
-    install_requires=["luma.core>=1.14.0"],
+    install_requires=["luma.core>=1.16.0"],
```
Comment (thijstriemstra) [inline setup.py:68]: this should be 1.16.1

**Review by rm-hull — APPROVED** [review]

File: setup.py (hunk 8 above)
Comment (thijstriemstra) [reply setup.py:68]: oh 1.16.1's not released yet though, we'll have to wait for that.

Comment (thijstriemstra) [reply setup.py]: fixed it.

**Comment by thijstriemstra** [comment]
@dhrone is luma.oled ready for a new release or should we hold off till these potential luma.core import changes?

**Comment by dhrone** [comment]
@thijstriemstra for some reason I thought this had already been published.  Since it has not, I'd hold up for a few days until I can make the changes to luma.core and also go through the documentation of both luma.lcd and luma.oled to make sure they are consistent.

**Pull request merged**